pkg/vere/main.c

```diff
@@ -1726,6 +1727,182 @@ _cw_prep(c3_i argc, c3_c* argv[])
+        c3_w lom_w;
+        c3_o res_o = _main_readw(optarg, u3a_bits + 3, &lom_w);
+        if ( (c3n == res_o) || (lom_w < 20) ) {
+          fprintf(stderr, "error: --loom must be >= 20 and <= %u\r\n", u3a_bits + 2);
+          u3_king_bail();
+        }
+        u3_Host.ops_u.lom_y = lom_w;
+      } break;
+
+      case '?': {
+        fprintf(stderr, "invalid argument\r\n");
+        u3_king_bail();
+      } break;
+    }
+  }
+
+  //  argv[optind] is always "chop"
+  //
+
+  if ( !u3_Host.dir_c ) {
+    if ( optind + 1 < argc ) {
+      u3_Host.dir_c = argv[optind + 1];
+    }
+    else {
+      fprintf(stderr, "invalid command, pier required\r\n");
+      u3_king_bail();
+    }
+
+    optind++;
+  }
+
+  if ( optind + 1 != argc ) {
+    fprintf(stderr, "invalid command\r\n");
+    u3_king_bail();
+  }
+
+  // gracefully shutdown the pier if it's running
+  u3_disk* old_u = _cw_disk_init(u3_Host.dir_c);
+
+  // note: this include patch applications (if any)
+  u3m_boot(u3_Host.dir_c, (size_t)1 << u3_Host.ops_u.lom_y);
+
+  // check if there's a *current* snapshot
+  if ( old_u->dun_d != u3A->eve_d ) {
+    fprintf(stderr, "chop: error: snapshot is out of date, please "
+                    "start/shutdown your pier gracefully first\r\n");
+    fprintf(stderr, "chop: eve_d: %" PRIu64 ", dun_d: %" PRIu64 "\r\n", u3A->eve_d, old_u->dun_d);
+    u3_king_bail();
+  }
+
+  if ( c3n == u3e_backup(c3y)) {  //  backup current snapshot
+    fprintf(stderr, "chop: error: failed to backup snapshot\r\n");
+    u3_king_bail();
+  }
+
+  // initialize the lmdb environment
+  // see disk.c:885
+  const size_t siz_i =
+  // 500 GiB is as large as musl on aarch64 wants to allow
+  #if (defined(U3_CPU_aarch64) && defined(U3_OS_linux))
+    0x7d00000000;
+  #else
+    0x10000000000;
+  #endif
+  c3_c log_c[8193];
+  snprintf(log_c, sizeof(log_c), "%s/.urb/log", u3_Host.dir_c);
+
+  // get the first/last event numbers from the event log
+  c3_d fir_d, las_d;
+  if ( c3n == u3_lmdb_gulf(old_u->mdb_u, &fir_d, &las_d) ) {
+    fprintf(stderr, "chop: failed to load latest event from database\r\n");
+    u3_king_bail();
+  }
+
+  // get the metadata
+  c3_d     who_d[2];
+  c3_o     fak_o;
+  c3_w     lif_w;
+  if ( c3y != u3_disk_read_meta(old_u->mdb_u, who_d, &fak_o, &lif_w) ) {
+    fprintf(stderr, "chop: failed to read metadata\r\n");
+    u3_king_bail();
+  }
+
+  // get the last event
+  u3_lmdb_walk itr_u;
+  size_t       len_i;
+  void*        buf_v[1];
+  if ( c3n == u3_lmdb_walk_init(old_u->mdb_u, &itr_u, las_d, las_d) ) {
+    fprintf(stderr, "chop: failed to initialize iterator\r\n");
+    u3_king_bail();
+  }
+  if ( c3n == u3_lmdb_walk_next(&itr_u, &len_i, buf_v) ) {
+    fprintf(stderr, "chop: failed to read event\r\n");
+    u3_king_bail();
+  }
+  u3_lmdb_walk_done(&itr_u);
+
+  // initialize a fresh lmdb environment in the "chop" subdir
+  c3_c cho_c[8193];
+  snprintf(cho_c, sizeof(cho_c), "%s/chop", log_c);
+  c3_mkdir(cho_c, 0700);
```

Comment:
The return code of `mkdir()` needs to be checked for errors as well.

Reply:
Thanks, WIP...